forte/processors/data_augment/algorithms/skeleton_op.py

```diff
@@ -0,0 +1,765 @@
+from forte.utils.utils import create_class_with_kwargs, get_class
+
+
+__all__ = ["SkeletonOp"]
+
+
+class SkeletonOp(ABC):
+    r"""
+    The SkeletonOp is the most basic augmentation Op that
+    gives users the most amount of freedom to implement their
+    logic of augmentation. The users are expected to use the
+    provided utility functions to implement their own augmentation
+    logic which will then be substantiated into new data packs. This
+    Op requires the users to have a relatively stronger understanding of
+    Forte's internal setup.
+    """
+
+    def __init__(self, configs: Union[Config, Dict[str, Any]]) -> None:
+
+        # :attr:`_replaced_annos`: {datapack id: SortedList[span, new text]}
+        # It records the spans replaced by new texts.
+        # It is a map from datapack id to a list of tuples
+        # (span, new text) inserted by :func:`replace`.
+        # The new text will be used for building new data pack.
+        self._replaced_annos: DefaultDict[
+            int, SortedList[Tuple[Span, str]]
+        ] = defaultdict(lambda: SortedList([], key=lambda x: x[0]))
+
+        # :attr:`_inserted_annos_pos_len`: {datapack id: Dict{position: length}}
+        # It records the position and length of inserted spans,
+        # mapping from datapack id to a dictionary (position -> length)
+        # inserted by :func:`insert`.
+        # The position is the index in the original datapack
+        # of insertion, and the length is the length of the inserted string.
+        self._inserted_annos_pos_len: DefaultDict[
+            int, SortedDict[int, int]
+        ] = defaultdict(lambda: SortedDict({}))
+
+        # :attr:`_deleted_annos_id`: {datapack id: Set[annotation tid]}
+        # It records the deleted span ids, mapping from datapack id
+        # to a set of annotation tids appended by :func:`delete`.
+        self._deleted_annos_id: DefaultDict[int, Set[int]] = defaultdict(set)
+
+        # :attr:`_data_pack_map`: {orig pack id: new pack id}
+        # It maintains a mapping from the pack id
+        # of the original pack to the pack id of augmented pack.
+        # It is used when copying the MultiPackLink and MultiPackGroup.
+        self._data_pack_map: Dict[int, int] = {}
+
+        # :attr:`_entry_maps`: {datapack id: Dict{orig tid, new tid}}
+        # It is a map for tracking the annotation ids
+        # before and after the auto align. It maps the
+        # original annotation tid to the new annotation tid.
+        # It is used when copying the Link/Group/MultiPackLink/MultiPackGroup.
+        self._entry_maps: Dict[int, Dict[int, int]] = {}
+
+        self.configs = configs
+
+    def perform_augmentation(self, input_pack: DataPack) -> DataPack:
+        r"""
+        Function to apply the defined augmentation function and
+        instantiating it into a data pack. This data pack is then
+        returned.
+
+        Args:
+            data_pack: The Datapack holding the replaced annotations.
```

Comment:
Have you try generating the docstring and take a look at the HTML? I feel the indent here is not correct.